crates/chia-datalayer/src/merkle.rs

```diff
@@ -0,0 +1,1897 @@
+        let mut base = match self {
+            Node::Internal(node) => node.to_bytes(),
+            Node::Leaf(node) => node.to_bytes(),
+        }
+        .map_err(Error::Streaming)?;
+        assert!(base.len() <= DATA_SIZE);
+        base.resize(DATA_SIZE, 0);
+        Ok(base
+            .as_slice()
+            .try_into()
+            .expect("padding was added above, might be too large"))
+    }
+
+    fn expect_leaf(&self, message: &str) -> LeafNode {
+        let Node::Leaf(leaf) = self else {
+            let message = message.replace("<<self>>", &format!("{self:?}"));
+            panic!("{}", message)
+        };
+
+        *leaf
+    }
+
+    fn try_into_leaf(self) -> Result<LeafNode, Error> {
+        match self {
+            Node::Leaf(leaf) => Ok(leaf),
+            Node::Internal(internal) => Err(Error::NodeNotALeaf(internal)),
+        }
+    }
+}
+
+#[cfg(feature = "py-bindings")]
+impl IntoPy<PyObject> for Node {
+    fn into_py(self, py: Python<'_>) -> PyObject {
+        match self {
+            Node::Internal(node) => node.into_py(py),
+            Node::Leaf(node) => node.into_py(py),
+        }
+    }
+}
+
+fn block_range(index: TreeIndex) -> Range<usize> {
+    let block_start = index.0 as usize * BLOCK_SIZE;
+    block_start..block_start + BLOCK_SIZE
+}
+
+pub struct Block {
+    // TODO: metadata node type and node's type not verified for agreement
+    metadata: NodeMetadata,
+    node: Node,
+}
+
+impl Block {
+    pub fn to_bytes(&self) -> Result<BlockBytes, Error> {
+        let mut blob: BlockBytes = [0; BLOCK_SIZE];
+        blob[METADATA_RANGE].copy_from_slice(&self.metadata.to_bytes().map_err(Error::Streaming)?);
+        blob[DATA_RANGE].copy_from_slice(&self.node.to_bytes()?);
+
+        Ok(blob)
+    }
+
+    pub fn from_bytes(blob: BlockBytes) -> Result<Self, Error> {
+        let metadata_blob: MetadataBytes = blob[METADATA_RANGE].try_into().unwrap();
+        let data_blob: DataBytes = blob[DATA_RANGE].try_into().unwrap();
+        let metadata = NodeMetadata::from_bytes(&metadata_blob)
+            .map_err(|message| Error::FailedLoadingMetadata(message.to_string()))?;
+        let node = Node::from_bytes(&metadata, &data_blob)
+            .map_err(|message| Error::FailedLoadingNode(message.to_string()))?;
+
+        Ok(Block { metadata, node })
+    }
+
+    pub fn update_hash(&mut self, left: &Hash, right: &Hash) {
+        self.node.set_hash(internal_hash(left, right));
+        self.metadata.dirty = false;
+    }
+}
+
+fn get_free_indexes_and_keys_values_indexes(
+    blob: &[u8],
+) -> (HashSet<TreeIndex>, HashMap<KvId, TreeIndex>) {
+    let index_count = blob.len() / BLOCK_SIZE;
+
+    let mut seen_indexes: Vec<bool> = vec![false; index_count];
+    let mut key_to_index: HashMap<KvId, TreeIndex> = HashMap::default();
+
+    for (index, block) in MerkleBlobLeftChildFirstIterator::new(blob) {
+        seen_indexes[index.0 as usize] = true;
```

Comment:
are we OK with a `panic!` here, if `index.0` is out of range?

Reply:
the all the indices being valid is an invariant that is maintained (unless there's a bug)